edb/edgeql/compiler/viewgen.py

```diff
@@ -1497,10 +1497,12 @@ def _normalize_view_ptr_expr(
         if irutils.is_trivial_free_object(ir_source):
             qlexpr = astutils.ensure_ql_select(qlexpr)
 
-        if ((ctx.expr_exposed or ctx.stmt is ctx.toplevel_stmt)
-                and ctx.implicit_limit
-                and isinstance(qlexpr, qlast.OffsetLimitMixin)
-                and not qlexpr.limit):
+        if (
+            (ctx.expr_exposed or ctx.stmt is ctx.toplevel_stmt)
+            and ctx.implicit_limit
+            and isinstance(qlexpr, (qlast.PipelinedQuery, qlast.ShapeElement))
```

Comment:
It looks like, starting with Python 3.10, it would actually work to keep saying `qlast.OffsetLimitMixin`! Maybe mypy doesn't believe that yet, though?

(Whether it actually works probably depends on the type. Maybe I need to make a mypy patch?)